### PR TITLE
Added query Enum value capability for vxlan EVPN feature

### PR DESCRIPTION
**Pull request by dgsudharsan**
Added query enum value capability for some attributes which are required for EVPN VxLAN feature.

### CODE REVIEW

**Review by kcudnik — DISMISSED** [review]

**Comment by kcudnik** [comment]
check build errors

**Review by kcudnik — APPROVED** [review]

**Comment by dgsudharsan** [comment]
> check build errors

@kcudnik @prsunny Can you please start pipeline?

**Comment by kcudnik** [comment]
this test failure is not related to the PR

**Comment by dgsudharsan** [comment]
> this test failure is not related to the PR

@kcudnik Can we merge this PR? What is the approach?

**Comment by kcudnik** [comment]
can you add unittests for this feature ? in sairedis/unittests/vslib/

File: vslib/SwitchStateBase.cpp [hunk 1]
```diff
@@ -3296,3 +3296,34 @@ bool SwitchStateBase::dumpObject(
...
+                              _In_ sai_object_id_t switch_id,
+                              _In_ sai_object_type_t object_type,
+                              _In_ sai_attr_id_t attr_id,
+                              _Inout_ sai_s32_list_t *enum_values_capability)
+{
+    SWSS_LOG_ENTER();
+
+    if (object_type == SAI_OBJECT_TYPE_TUNNEL && attr_id == SAI_TUNNEL_ATTR_PEER_MODE)
+    {
+        return queryVxlanTunnelPeerModeCapability(enum_values_capability);
```
Comment (kcudnik) [inline vslib/SwitchStateBase.cpp:3326]: why there is Vxlan in in name here? there is no vxlan object type nor attribute in name, this is just for tunnel

**Comment by dgsudharsan** [comment]
> sairedis/unittests/vslib/

Sure. Can you please share the steps on how to run these unit tests?

**Comment by kcudnik** [comment]
yea, just type "make check" in test directory

File: vslib/SwitchStateBase.cpp (hunk 1 above)
Comment (dgsudharsan) [reply vslib/SwitchStateBase.cpp]: Removed

**Comment by dgsudharsan** [comment]
> yea, just type "make check" in test directory

I ran the unit tests

[----------] 5 tests from SwitchMLNX2700
[ RUN      ] SwitchMLNX2700.ctr
[       OK ] SwitchMLNX2700.ctr (14 ms)
[ RUN      ] SwitchMLNX2700.refresh_bridge_port_list
[       OK ] SwitchMLNX2700.refresh_bridge_port_list (10 ms)
[ RUN      ] SwitchMLNX2700.warm_update_queues
[       OK ] SwitchMLNX2700.warm_update_queues (17 ms)
[ RUN      ] SwitchMLNX2700.test_tunnel_term_capability
[       OK ] SwitchMLNX2700.test_tunnel_term_capability (4 ms)
[ RUN      ] SwitchMLNX2700.test_vlan_flood_capability
[       OK ] SwitchMLNX2700.test_vlan_flood_capability (4 ms)
[----------] 5 tests from SwitchMLNX2700 (49 ms total)

[----------] 5 tests from SwitchBCM56850
[ RUN      ] SwitchBCM56850.ctr
[       OK ] SwitchBCM56850.ctr (16 ms)
[ RUN      ] SwitchBCM56850.refresh_bridge_port_list
[       OK ] SwitchBCM56850.refresh_bridge_port_list (11 ms)
[ RUN      ] SwitchBCM56850.warm_update_queues
[       OK ] SwitchBCM56850.warm_update_queues (20 ms)
[ RUN      ] SwitchBCM56850.test_tunnel_term_capability
[       OK ] SwitchBCM56850.test_tunnel_term_capability (5 ms)
[ RUN      ] SwitchBCM56850.test_vlan_flood_capability
[       OK ] SwitchBCM56850.test_vlan_flood_capability (4 ms)
[----------] 5 tests from SwitchBCM56850 (56 ms total)

**Comment by dgsudharsan** [comment]
@kcudnik I have added tests to validate new APIs
However, the infrastructure itself 'works' outside sai library. If fake_platform isset it will choose appropriate saixml file. The infrastructure is tested in swss tests which make use of the fake_platform concept. https://github.com/Azure/sonic-swss/pull/1858

**Comment by kcudnik** [comment]
fake_platfrom sholud not be used anywhere its a terrible idea to use that variable anywhere

**Comment by dgsudharsan** [comment]
> fake_platfrom sholud not be used anywhere its a terrible idea to use that variable anywhere

The infrastructure cleanup it's a big work item that will be done in some later point of time. For now fake_platform isn't used anywhere inside sai-redis
Please let me know if the current changes are good. If so can you please approve this? We have a feature pending on this.

**Comment by kcudnik** [comment]
sonic project is a classic book example for "will be done in some later point" translates to never

File: unittest/vslib/TestSwitchBCM56850.cpp [hunk 2]
```diff
@@ -268,3 +268,135 @@ TEST(SwitchBCM56850, warm_update_queues)
...
+            0x2100000000,
+            std::make_shared<RealObjectIdManager>(0, scc),
+            sc);
+
+
+    sai_s32_list_t enum_val_cap;
+    int32_t list[2];
+    enum_val_cap.count = 2;
+    enum_val_cap.list = list;
+    EXPECT_EQ(sw.queryAttrEnumValuesCapability(0x2100000000, SAI_OBJECT_TYPE_TUNNEL,
```
Comment (kcudnik) [inline unittest/vslib/TestSwitchBCM56850.cpp:299]: please use 1 parameter per line if you want to break function params to multiple lines, also do this across this PR

File: unittest/vslib/TestSwitchBCM56850.cpp [hunk 3]
```diff
@@ -268,3 +268,135 @@ TEST(SwitchBCM56850, warm_update_queues)
...
+                                               SAI_TUNNEL_ATTR_PEER_MODE, &enum_val_cap),
+                                               SAI_STATUS_SUCCESS);
+
+    EXPECT_EQ(enum_val_cap.count, 2);
+
+    int modes_found = 0;
+    for (uint32_t i = 0; i < enum_val_cap.count; i++)
```
Comment (kcudnik) [inline unittest/vslib/TestSwitchBCM56850.cpp:306]: add empty line before for

File: unittest/vslib/TestSwitchBCM56850.cpp [hunk 4]
```diff
@@ -268,3 +268,135 @@ TEST(SwitchBCM56850, warm_update_queues)
...
+    {
+        if (enum_val_cap.list[i] == SAI_TUNNEL_PEER_MODE_P2MP || enum_val_cap.list[i] == SAI_TUNNEL_PEER_MODE_P2P)
+        {
+            modes_found++;
+        }
+    }
+    EXPECT_EQ(modes_found, 2);
+
```
Comment (kcudnik) [inline unittest/vslib/TestSwitchBCM56850.cpp:314]: remove empty line here

File: unittest/vslib/TestSwitchBCM56850.cpp [hunk 5]
```diff
@@ -268,3 +268,135 @@ TEST(SwitchBCM56850, warm_update_queues)
...
+    EXPECT_EQ(enum_val_cap.count, 2);
+
+    int modes_found = 0;
+    for (uint32_t i = 0; i < enum_val_cap.count; i++)
+    {
+        if (enum_val_cap.list[i] == SAI_TUNNEL_PEER_MODE_P2MP || enum_val_cap.list[i] == SAI_TUNNEL_PEER_MODE_P2P)
+        {
+            modes_found++;
+        }
+    }
```
Comment (kcudnik) [inline unittest/vslib/TestSwitchBCM56850.cpp:312]: add empty line after } and another code

File: unittest/vslib/TestSwitchBCM56850.cpp (hunk 3 above)
Comment (dgsudharsan) [reply unittest/vslib/TestSwitchBCM56850.cpp:309]: Done

File: unittest/vslib/TestSwitchBCM56850.cpp (hunk 5 above)
Comment (dgsudharsan) [reply unittest/vslib/TestSwitchBCM56850.cpp:315]: Done

File: unittest/vslib/TestSwitchBCM56850.cpp (hunk 2 above)
Comment (dgsudharsan) [reply unittest/vslib/TestSwitchBCM56850.cpp]: Done

**Comment by dgsudharsan** [comment]
> sonic project is a classic book example for "will be done in some later point" translates to never

Taking up the entire cleanup work is not within the timelines of the current feature. It's a new work item by itself. We have prioritized this work item as it is also required for us to specify platform and I agree fake_platform isn't a good approach.

File: unittest/vslib/TestSwitchBCM56850.cpp (hunk 4 above)
Comment (dgsudharsan) [reply unittest/vslib/TestSwitchBCM56850.cpp:318]: Done

**Review by kcudnik — APPROVED** [review]

**Comment by dgsudharsan** [comment]
The test failures are not related to changes in the PR

**Pull request merged**